libraries/chain/controller.cpp

```diff
@@ -3898,6 +3897,8 @@ struct controller_impl {
          consider_voting(bsp, use_thread_pool_t::no);
       }
 
+      // should_terminate() is not thread-safe requires chain_head.block_num()
+      // also when syncing allow many blocks in forkdb queued to be applied
```

Comment:
What about having both:

```
// thread-safe
bool should_terminate(block_num_type head_block_num) const {
   if (conf.terminate_at_block > 0 && conf.terminate_at_block <= head_block_num) {
      ilog("Block ${n} reached configured maximum block ${num}; terminating",
           ("n", head_block_num)("num", conf.terminate_at_block) );
      return true;
   }
   return false;
}

// not thread-safe because of chain_head access
bool should_terminate() const {
   return should_terminate(chain_head.block_num());
}
```